File: demo-app/app/src/main/java/com/launchkey/android/authenticator/demo/ui/adapter/DemoSessionsAdapter.java
package com.launchkey.android.authenticator.demo.ui.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.TextView;

import com.launchkey.android.authenticator.sdk.session.Session;
import com.launchkey.android.authenticator.demo.R;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Created by armando on 7/11/16.
 */
public class DemoSessionsAdapter extends BaseAdapter {

    private Context mContext;
    private List<Session> mSessions = new ArrayList<>();
    private AdapterView.OnItemClickListener mItemClickListener;
    private View.OnClickListener mInternalClickListener;

    public DemoSessionsAdapter(Context c, List<Session> sessions, AdapterView.OnItemClickListener l) {
        mContext = c;
        mSessions = sessions;
        mItemClickListener = l;

        mInternalClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (v != null && v.getTag() != null) {
                    int position = (int) v.getTag();

                    if (mItemClickListener != null) {
                        mItemClickListener.onItemClick(null, v, position, position);
                    }
                }
            }
        };
    }

    @Override
    public int getCount() {
        return mSessions.size();
    }

    @Override
    public Session getItem(int position) {
        return mSessions.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        //Relying on the authorizations layout for items

        View v = convertView;

        if (v == null) {
            v = LayoutInflater
                    .from(mContext)
                    .inflate(R.layout.demo_authorizations_item, parent, false);
        }

        Session s = getItem(position);

        //fetch an application's icon via its url with Application.getAppIcon()

        TextView name = (TextView) v.findViewById(R.id.demo_authorizations_item_name);
        name.setText(s.getName());

        //TODO: Update usage of ID as placeholder for potential context being sent
        TextView context = (TextView) v.findViewById(R.id.demo_authorizations_item_context);
        context.setText(s.getId());

        long millisAgo = s.getCreatedAgoMillis(mContext);

        TextView action = (TextView) v.findViewById(R.id.demo_authorizations_item_text_action);
        action.setText(String.format(Locale.getDefault(), "%d seconds ago", (millisAgo / 1000)));

        /*
        TextView status = (TextView) v.findViewById(R.id.demo_authorizations_item_text_status);
        status.setText(a.getStatus().toUpperCase());

        TextView transactional = (TextView) v.findViewById(R.id.demo_authorizations_item_text_transactional);
        transactional.setText(a.isTransactional() ? "TRANSACTIONAL" : "");
        */

        Button button = (Button) v.findViewById(R.id.demo_authorizations_item_button);
        button.setText("LOG OUT");
        button.setOnClickListener(mInternalClickListener);
        button.setTag(position);

        return v;
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/adapter/DemoSessionsAdapter.kt
package com.launchkey.android.authenticator.demo.ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.TextView
import com.launchkey.android.authenticator.demo.R
import com.launchkey.android.authenticator.sdk.session.Session
import java.util.*

/**
 * Created by armando on 7/11/16.
 */
class DemoSessionsAdapter(private val mContext: Context, sessions: List<Session>, private val mItemClickListener: AdapterView.OnItemClickListener?) : BaseAdapter() {
    private var mSessions: List<Session> = ArrayList<Session>()
    private val mInternalClickListener: View.OnClickListener

    init {
        mSessions = sessions

        mInternalClickListener = View.OnClickListener { v ->
            if (v != null && v.tag != null) {
                val position = v.tag as Int

                mItemClickListener?.onItemClick(null, v, position, position.toLong())
            }
        }
    }

    override fun getCount(): Int {
        return mSessions.size
    }

    override fun getItem(position: Int): Session {
        return mSessions[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View, parent: ViewGroup): View {

        //Relying on the authorizations layout for items

        var v: View? = convertView

        if (v == null) {
            v = LayoutInflater
                    .from(mContext)
                    .inflate(R.layout.demo_authorizations_item, parent, false)
        }

        val s = getItem(position)

        //fetch an application's icon via its url with Application.getAppIcon()

        val name = v!!.findViewById<View>(R.id.demo_authorizations_item_name) as TextView
        name.text = s.name

        //TODO: Update usage of ID as placeholder for potential context being sent
        val context = v.findViewById<View>(R.id.demo_authorizations_item_context) as TextView
        context.text = s.id

        val millisAgo = s.getCreatedAgoMillis(mContext)

        val action = v.findViewById<View>(R.id.demo_authorizations_item_text_action) as TextView
        action.text = String.format(Locale.getDefault(), "%d seconds ago", millisAgo / 1000)

        /*
        TextView status = (TextView) v.findViewById(R.id.demo_authorizations_item_text_status);
        status.setText(a.getStatus().toUpperCase());

        TextView transactional = (TextView) v.findViewById(R.id.demo_authorizations_item_text_transactional);
        transactional.setText(a.isTransactional() ? "TRANSACTIONAL" : "");
        */

        val button = v.findViewById<View>(R.id.demo_authorizations_item_button) as Button
        button.text = "LOG OUT"
        button.setOnClickListener(mInternalClickListener)
        button.tag = position

        return v
    }
}
